src/main/java/org/osc/manager/ism/api/IsmDeviceApi.java

```diff
@@ -138,80 +181,137 @@ public Void call() throws Exception {
+
             @Override
-            public DeviceMember call() throws Exception {
-                VSSDevice device = em.find(VSSDevice.class, vs.getId());
-                
-                DeviceMember dm = new DeviceMember();
-                dm.setParent(device);
-                dm.setName(name);
-                em.persist(dm);
-                return dm;
+            public DeviceMemberEntity call() throws Exception {
+
+                CriteriaBuilder criteriaBuilder = IsmDeviceApi.this.em.getCriteriaBuilder();
+                CriteriaQuery<DeviceEntity> query = criteriaBuilder.createQuery(DeviceEntity.class);
+                Root<DeviceEntity> r = query.from(DeviceEntity.class);
+                query.select(r)
+                        .where(criteriaBuilder.and(criteriaBuilder.equal(r.get("vsId"), IsmDeviceApi.this.vs.getId())));
+                List<DeviceEntity> result = IsmDeviceApi.this.em.createQuery(query).getResultList();
+
+                if (result.isEmpty()) {
+                    throw new Exception("Device Entity does not exists...");
+                    //TODO - to add RETURN 404 error:Sudhir
+                }
+                CriteriaBuilder memberBuilder = IsmDeviceApi.this.em.getCriteriaBuilder();
+                CriteriaQuery<DeviceMemberEntity> memberQuery = memberBuilder.createQuery(DeviceMemberEntity.class);
+                Root<DeviceMemberEntity> res = memberQuery.from(DeviceMemberEntity.class);
+                memberQuery.select(res).where(memberBuilder.and(memberBuilder.equal(r.get("name"), name)));
+
+                List<DeviceMemberEntity> deviceMemberResult = IsmDeviceApi.this.em.createQuery(memberQuery)
+                        .getResultList();
+                if (!deviceMemberResult.isEmpty()) {
+                    throw new Exception("Device Member Entity name already exists...:");
+                    //TODO - to add RETURN 404 error:Sudhir
+                }
+                DeviceMemberEntity entity = new DeviceMemberEntity();
+                entity.setParent(result.get(0));
+                entity.setName(name);
+                IsmDeviceApi.this.em.persist(entity);
+                return entity;
             }
-        }).getId().toString();
+        }).getId();
     }
 
 
     @Override
     public String updateDeviceMember(ManagerDeviceMemberElement deviceElement, String name, String vserverIpAddress,
             String contactIpAddress, String ipAddress, String gateway, String prefixLength) throws Exception {
-        return deviceElement.getId();
+        return this.txControl.required(new Callable<DeviceMemberEntity>() {
+
+            @Override
+            public DeviceMemberEntity call() throws Exception {
+
+                CriteriaBuilder criteriaBuilder = IsmDeviceApi.this.em.getCriteriaBuilder();
+
+                CriteriaQuery<DeviceMemberEntity> query = criteriaBuilder.createQuery(DeviceMemberEntity.class);
```

Comment:
I think you need to query the parent device and filter the specific device member from its children by id. This method is currently not taking into consideration the parent device. Also, similarly to my other comment I can imagine you can have a common method to updateDeviceMember that takes the memberEntity as parameter. This api method can then construct this entity and call that method, just like create. The rest api can do the same.  You common code to deal with the DB always use entities, the SDK implementation can construct those entities using the vs info or other parameters (like in this case deviceElement, name, etc..) and call that common code.  So the patter is API/SDK implementation -> creates the entity to be updated, deleted, etc -> calls common code/method, rest api -> calls common code. Everywhere where there is parity of operations this should apply. If something is just in the rest api, i.e.: some delete operation it does not need to be common.

Reply:
Regarding the remark "I think you need to query the parent device and filter the specific device member from its children by id. This method is currently not taking into consideration the parent device" ==> This is not needed, because I am querying on specific unique devicemember ID and should be good enough.

Reply:
OK, I have taken care.... in some cases query is based on vsId versus Id and have refactored this in the common method.

Reply:
This is partially true. This method assumes the device returned is associated with the vs id of the parent (remember the vs is passed to this api). . Say someone is querying a device by id that does not belong to the targeted vs id (parent device with that vs id). This method will return the device member because the member id is found in the db but it should not have since the parent did not match. This method should take two additional paremeters, the parent vcId and the parent id. When called by the rest layer it receives the parent id, when called by the SDK api implemented it receives the vs id. It queries the parent by either one of those and ensure it exists. The bottom line is, you should not update, delete or create any child entity without ensuring the parent exists.